File: LeetCode/Medium/0143-reorder-list/0143-reorder-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
    // Edge case: If the list is empty or has only one node, no reordering is needed
    if head == nil || head.Next == nil {
        return
    }
    
    // Step 1: Find the middle of the linked list
    slow, fast := head, head.Next
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    
    // Step 2: Reverse the second half
    second := slow.Next
    slow.Next = nil // Split the list into two separate halves
    
    var prev *ListNode // In Go, pointers default to nil
    for second != nil {
        tmp := second.Next
        second.Next = prev
        prev = second
        second = tmp
    }
    
    // Step 3: Merge the two halves
    first := head
    second = prev // 'prev' is the head of the reversed half
    
    for second != nil {
        tmp1, tmp2 := first.Next, second.Next
        
        // Link nodes alternately
        first.Next = second
        second.Next = tmp1
        
        // Move pointers forward
        first = tmp1
        second = tmp2
    }
}

File: LeetCode/Medium/0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # Edge case: If the list is empty or has only one node, no reordering is needed
        if not head or not head.next:
            return
        
        # first reach the mid point of the linkedlist
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # now reverse the second half of the linkedlist
        second = slow.next
        slow.next = None
        prev = None
        while second:
            next_node = second.next
            second.next = prev
            prev = second
            second = next_node
        second = prev

        # now merge the two halfs of the linkedList
        first = head
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

        return head
        